File: js/src/format.ts
/*
 * Internal: MessagePack format constants, byte-order helpers, the growable
 * output buffer, and skip_one. Private to the implementation; mirrors
 * cpp/src/msgpack_blob_detail.hpp and the skip routine from the decode module.
 */

export const MAX_DEPTH = 200;
export const MAX_OUTPUT = 64 * 1024 * 1024;

export const MP_NIL = 0xc0;
export const MP_FALSE = 0xc2;
export const MP_TRUE = 0xc3;
export const MP_BIN8 = 0xc4;
export const MP_BIN16 = 0xc5;
export const MP_BIN32 = 0xc6;
export const MP_EXT8 = 0xc7;
export const MP_EXT16 = 0xc8;
export const MP_EXT32 = 0xc9;
export const MP_FLOAT32 = 0xca;
export const MP_FLOAT64 = 0xcb;
export const MP_UINT8 = 0xcc;
export const MP_UINT16 = 0xcd;
export const MP_UINT32 = 0xce;
export const MP_UINT64 = 0xcf;
export const MP_INT8 = 0xd0;
export const MP_INT16 = 0xd1;
export const MP_INT32 = 0xd2;
export const MP_INT64 = 0xd3;
export const MP_FIXEXT1 = 0xd4;
export const MP_FIXEXT2 = 0xd5;
export const MP_FIXEXT4 = 0xd6;
export const MP_FIXEXT8 = 0xd7;
export const MP_FIXEXT16 = 0xd8;
export const MP_STR8 = 0xd9;
export const MP_STR16 = 0xda;
export const MP_STR32 = 0xdb;
export const MP_ARRAY16 = 0xdc;
export const MP_ARRAY32 = 0xdd;
export const MP_MAP16 = 0xde;
export const MP_MAP32 = 0xdf;

export const MP_FIXMAP_MASK = 0x80;
export const MP_FIXARRAY_MASK = 0x90;
export const MP_FIXSTR_MASK = 0xa0;

export const MP_TIMESTAMP_TYPE = 0xff;

export const U64_MASK = 0xffffffffffffffffn;

// ── big-endian read helpers ─────────────────────────────────────────
export function read16(a: Uint8Array, i: number): number {
  return (a[i] << 8) | a[i + 1];
}
export function read32(a: Uint8Array, i: number): number {
  return ((a[i] << 24) | (a[i + 1] << 16) | (a[i + 2] << 8) | a[i + 3]) >>> 0;
}
export function read64(a: Uint8Array, i: number): bigint {
  let v = 0n;
  for (let k = 0; k < 8; k++) v = (v << 8n) | BigInt(a[i + k]);
  return v;
}

const _scratch = new DataView(new ArrayBuffer(8));
export function readF32(a: Uint8Array, i: number): number {
  for (let k = 0; k < 4; k++) _scratch.setUint8(k, a[i + k]);
  return _scratch.getFloat32(0, false);
}
export function readF64(a: Uint8Array, i: number): number {
  for (let k = 0; k < 8; k++) _scratch.setUint8(k, a[i + k]);
  return _scratch.getFloat64(0, false);
}

// ── growable output buffer ──────────────────────────────────────────
export class Buf {
  bytes: number[];
  constructor() {
    this.bytes = [];
  }
  get length(): number {
    return this.bytes.length;
  }
  push(b: number): void {
    this.bytes.push(b & 0xff);
  }
  pushU16(v: number): void {
    this.bytes.push((v >>> 8) & 0xff, v & 0xff);
  }
  pushU32(v: number): void {
    this.bytes.push((v >>> 24) & 0xff, (v >>> 16) & 0xff, (v >>> 8) & 0xff, v & 0xff);
  }
  pushU64(v: bigint): void {
    const b = BigInt.asUintN(64, v);
    for (let s = 56n; s >= 0n; s -= 8n) this.bytes.push(Number((b >> s) & 0xffn));
  }
  pushF32(f: number): void {
    _scratch.setFloat32(0, f, false);
    for (let k = 0; k < 4; k++) this.bytes.push(_scratch.getUint8(k));
  }
  pushF64(d: number): void {
    _scratch.setFloat64(0, d, false);
    for (let k = 0; k < 8; k++) this.bytes.push(_scratch.getUint8(k));
  }
  pushBytes(src: Uint8Array | number[]): void {
    for (let k = 0; k < src.length; k++) this.bytes.push(src[k] & 0xff);
  }
  pushSlice(a: Uint8Array, start: number, end: number): void {
    for (let i = start; i < end; i++) this.bytes.push(a[i]);
  }
  toBytes(): Uint8Array {
    return Uint8Array.from(this.bytes);
  }
}

// ── skip_one — offset just past one element, or 0 on malformed input ─
export function skipOne(a: Uint8Array, n: number, i: number): number {
  return skipOneD(a, n, i, 0);
}

function skipOneD(a: Uint8Array, n: number, i: number, depth: number): number {
  if (depth > MAX_DEPTH) return 0;
  if (i >= n) return 0;
  const b = a[i];
  i += 1;

  if (b <= 0x7f) return i;
  if (b >= 0xe0) return i;

  switch (b) {
    case MP_NIL:
    case MP_FALSE:
    case MP_TRUE:
      return i;
    case MP_FLOAT32:
      return i + 4 <= n ? i + 4 : 0;
    case MP_FLOAT64:
    case MP_INT64:
    case MP_UINT64:
      return i + 8 <= n ? i + 8 : 0;
    case MP_UINT8:
    case MP_INT8:
      return i + 1 <= n ? i + 1 : 0;
    case MP_UINT16:
    case MP_INT16:
      return i + 2 <= n ? i + 2 : 0;
    case MP_UINT32:
    case MP_INT32:
      return i + 4 <= n ? i + 4 : 0;
    case MP_BIN8:
    case MP_STR8: {
      if (i + 1 > n) return 0;
      const sz = a[i];
      i += 1;
      return sz <= n - i ? i + sz : 0;
    }
    case MP_BIN16:
    case MP_STR16: {
      if (i + 2 > n) return 0;
      const sz = read16(a, i);
      i += 2;
      return sz <= n - i ? i + sz : 0;
    }
    case MP_BIN32:
    case MP_STR32: {
      if (i + 4 > n) return 0;
      const sz = read32(a, i);
      i += 4;
      return sz <= n - i ? i + sz : 0;
    }
    case MP_FIXEXT1:
      return i + 2 <= n ? i + 2 : 0;
    case MP_FIXEXT2:
      return i + 3 <= n ? i + 3 : 0;
    case MP_FIXEXT4:
      return i + 5 <= n ? i + 5 : 0;
    case MP_FIXEXT8:
      return i + 9 <= n ? i + 9 : 0;
    case MP_FIXEXT16:
      return i + 17 <= n ? i + 17 : 0;
    case MP_EXT8: {
      if (i + 2 > n) return 0;
      const sz = a[i];
      i += 2;
      return sz <= n - i ? i + sz : 0;
    }
    case MP_EXT16: {
      if (i + 3 > n) return 0;
      const sz = read16(a, i);
      i += 3;
      return sz <= n - i ? i + sz : 0;
    }
    case MP_EXT32: {
      if (i + 5 > n) return 0;
      const sz = read32(a, i);
      i += 5;
      return sz <= n - i ? i + sz : 0;
    }
    default:
      break;
  }

  if (b >= 0xa0 && b <= 0xbf) {
    const sz = b & 0x1f;
    return sz <= n - i ? i + sz : 0;
  }

  if (b >= 0x90 && b <= 0x9f) {
    const count = b & 0x0f;
    for (let j = 0; j < count; j++) {
      i = skipOneD(a, n, i, depth + 1);
      if (!i) return 0;
    }
    return i;
  }

  if (b >= 0x80 && b <= 0x8f) {
    const count = b & 0x0f;
    for (let j = 0; j < count; j++) {
      i = skipOneD(a, n, i, depth + 1);
      if (!i) return 0;
      i = skipOneD(a, n, i, depth + 1);
      if (!i) return 0;
    }
    return i;
  }

  if (b === MP_ARRAY16 || b === MP_ARRAY32) {
    let count: number;
    if (b === MP_ARRAY16) {
      if (i + 2 > n) return 0;
      count = read16(a, i);
      i += 2;
    } else {
      if (i + 4 > n) return 0;
      count = read32(a, i);
      i += 4;
    }
    for (let j = 0; j < count; j++) {
      i = skipOneD(a, n, i, depth + 1);
      if (!i) return 0;
    }
    return i;
  }

  if (b === MP_MAP16 || b === MP_MAP32) {
    let count: number;
    if (b === MP_MAP16) {
      if (i + 2 > n) return 0;
      count = read16(a, i);
      i += 2;
    } else {
      if (i + 4 > n) return 0;
      count = read32(a, i);
      i += 4;
    }
    for (let j = 0; j < count; j++) {
      i = skipOneD(a, n, i, depth + 1);
      if (!i) return 0;
      i = skipOneD(a, n, i, depth + 1);
      if (!i) return 0;
    }
    return i;
  }

  return 0;
}

// ── UTF-8 helpers ───────────────────────────────────────────────────
// Byte-preserving (lossless) UTF-8 codec matching Python's
// `bytes.decode("utf-8", "surrogateescape")` / `str.encode(...)`. The C++
// reference passes raw `str` bytes through verbatim, so a plain replacing
// TextDecoder (which substitutes U+FFFD) would diverge on non-UTF-8 input.
// Invalid bytes round-trip through lone low surrogates U+DC80..U+DCFF.
export function utf8Decode(bytes: Uint8Array): string {
  let out = "";
  let i = 0;
  const n = bytes.length;
  while (i < n) {
    const b0 = bytes[i];
    if (b0 < 0x80) {
      out += String.fromCharCode(b0);
      i++;
      continue;
    }
    let len: number;
    let cp: number;
    if (b0 >= 0xc2 && b0 <= 0xdf) {
      len = 2;
      cp = b0 & 0x1f;
    } else if (b0 >= 0xe0 && b0 <= 0xef) {
      len = 3;
      cp = b0 & 0x0f;
    } else if (b0 >= 0xf0 && b0 <= 0xf4) {
      len = 4;
      cp = b0 & 0x07;
    } else {
      out += String.fromCharCode(0xdc00 + b0);
      i++;
      continue;
    }
    if (i + len > n) {
      out += String.fromCharCode(0xdc00 + b0);
      i++;
      continue;
    }
    const b1 = bytes[i + 1];
    let valid: boolean;
    if (len === 2) valid = b1 >= 0x80 && b1 <= 0xbf;
    else if (len === 3) {
      if (b0 === 0xe0) valid = b1 >= 0xa0 && b1 <= 0xbf;
      else if (b0 === 0xed) valid = b1 >= 0x80 && b1 <= 0x9f; // exclude surrogates
      else valid = b1 >= 0x80 && b1 <= 0xbf;
    } else {
      if (b0 === 0xf0) valid = b1 >= 0x90 && b1 <= 0xbf;
      else if (b0 === 0xf4) valid = b1 >= 0x80 && b1 <= 0x8f;
      else valid = b1 >= 0x80 && b1 <= 0xbf;
    }
    if (!valid) {
      out += String.fromCharCode(0xdc00 + b0);
      i++;
      continue;
    }
    cp = (cp << 6) | (b1 & 0x3f);
    let ok = true;
    for (let k = 2; k < len; k++) {
      const bk = bytes[i + k];
      if (bk < 0x80 || bk > 0xbf) {
        ok = false;
        break;
      }
      cp = (cp << 6) | (bk & 0x3f);
    }
    if (!ok) {
      out += String.fromCharCode(0xdc00 + b0);
      i++;
      continue;
    }
    if (cp <= 0xffff) out += String.fromCharCode(cp);
    else {
      cp -= 0x10000;
      out += String.fromCharCode(0xd800 + (cp >> 10), 0xdc00 + (cp & 0x3ff));
    }
    i += len;
  }
  return out;
}

export function utf8Encode(s: string): Uint8Array {
  const out: number[] = [];
  for (let i = 0; i < s.length; i++) {
    let cp = s.charCodeAt(i);
    if (cp >= 0xdc80 && cp <= 0xdcff) {
      out.push(cp & 0xff); // surrogateescape byte
      continue;
    }
    if (cp >= 0xd800 && cp <= 0xdbff && i + 1 < s.length) {
      const lo = s.charCodeAt(i + 1);
      if (lo >= 0xdc00 && lo <= 0xdfff) {
        cp = 0x10000 + ((cp - 0xd800) << 10) + (lo - 0xdc00);
        i++;
      }
    }
    if (cp < 0x80) out.push(cp);
    else if (cp < 0x800) out.push(0xc0 | (cp >> 6), 0x80 | (cp & 0x3f));
    else if (cp < 0x10000) out.push(0xe0 | (cp >> 12), 0x80 | ((cp >> 6) & 0x3f), 0x80 | (cp & 0x3f));
    else out.push(0xf0 | (cp >> 18), 0x80 | ((cp >> 12) & 0x3f), 0x80 | ((cp >> 6) & 0x3f), 0x80 | (cp & 0x3f));
  }
  return Uint8Array.from(out);
}

File: rust/src/format.rs
//! Internal: MessagePack format constants, byte-order helpers and `skip_one`.
//!
//! Private to the crate; mirrors `cpp/src/msgpack_blob_detail.hpp` and the skip
//! routine from the C++ decode module.

pub const MAX_DEPTH: i32 = 200;
pub const MAX_OUTPUT: usize = 64 * 1024 * 1024;

pub const MP_NIL: u8 = 0xc0;
pub const MP_FALSE: u8 = 0xc2;
pub const MP_TRUE: u8 = 0xc3;
pub const MP_BIN8: u8 = 0xc4;
pub const MP_BIN16: u8 = 0xc5;
pub const MP_BIN32: u8 = 0xc6;
pub const MP_EXT8: u8 = 0xc7;
pub const MP_EXT16: u8 = 0xc8;
pub const MP_EXT32: u8 = 0xc9;
pub const MP_FLOAT32: u8 = 0xca;
pub const MP_FLOAT64: u8 = 0xcb;
pub const MP_UINT8: u8 = 0xcc;
pub const MP_UINT16: u8 = 0xcd;
pub const MP_UINT32: u8 = 0xce;
pub const MP_UINT64: u8 = 0xcf;
pub const MP_INT8: u8 = 0xd0;
pub const MP_INT16: u8 = 0xd1;
pub const MP_INT32: u8 = 0xd2;
pub const MP_INT64: u8 = 0xd3;
pub const MP_FIXEXT1: u8 = 0xd4;
pub const MP_FIXEXT2: u8 = 0xd5;
pub const MP_FIXEXT4: u8 = 0xd6;
pub const MP_FIXEXT8: u8 = 0xd7;
pub const MP_FIXEXT16: u8 = 0xd8;
pub const MP_STR8: u8 = 0xd9;
pub const MP_STR16: u8 = 0xda;
pub const MP_STR32: u8 = 0xdb;
pub const MP_ARRAY16: u8 = 0xdc;
pub const MP_ARRAY32: u8 = 0xdd;
pub const MP_MAP16: u8 = 0xde;
pub const MP_MAP32: u8 = 0xdf;

pub const MP_FIXMAP_MASK: u8 = 0x80;
pub const MP_FIXARRAY_MASK: u8 = 0x90;
pub const MP_FIXSTR_MASK: u8 = 0xa0;

pub const MP_TIMESTAMP_TYPE: u8 = 0xff;

// ── big-endian read helpers ─────────────────────────────────────────
#[inline]
pub fn read16(a: &[u8], i: usize) -> u32 {
    ((a[i] as u32) << 8) | a[i + 1] as u32
}
#[inline]
pub fn read32(a: &[u8], i: usize) -> u32 {
    ((a[i] as u32) << 24) | ((a[i + 1] as u32) << 16) | ((a[i + 2] as u32) << 8) | a[i + 3] as u32
}
#[inline]
pub fn read64(a: &[u8], i: usize) -> u64 {
    ((read32(a, i) as u64) << 32) | read32(a, i + 4) as u64
}

// ── big-endian write helpers ────────────────────────────────────────
#[inline]
pub fn push16(out: &mut Vec<u8>, v: u16) {
    out.extend_from_slice(&v.to_be_bytes());
}
#[inline]
pub fn push32(out: &mut Vec<u8>, v: u32) {
    out.extend_from_slice(&v.to_be_bytes());
}
#[inline]
pub fn push64(out: &mut Vec<u8>, v: u64) {
    out.extend_from_slice(&v.to_be_bytes());
}

/// Return the offset just past one complete element starting at `i`,
/// or 0 on malformed / truncated input.
pub fn skip_one(a: &[u8], n: usize, i: usize) -> usize {
    skip_one_d(a, n, i, 0)
}

fn skip_one_d(a: &[u8], n: usize, mut i: usize, depth: i32) -> usize {
    if depth > MAX_DEPTH {
        return 0;
    }
    if i >= n {
        return 0;
    }
    let b = a[i];
    i += 1;

    if b <= 0x7f {
        return i;
    }
    if b >= 0xe0 {
        return i;
    }

    match b {
        MP_NIL | MP_FALSE | MP_TRUE => return i,
        MP_FLOAT32 => return if i + 4 <= n { i + 4 } else { 0 },
        MP_FLOAT64 | MP_INT64 | MP_UINT64 => return if i + 8 <= n { i + 8 } else { 0 },
        MP_UINT8 | MP_INT8 => return if i + 1 <= n { i + 1 } else { 0 },
        MP_UINT16 | MP_INT16 => return if i + 2 <= n { i + 2 } else { 0 },
        MP_UINT32 | MP_INT32 => return if i + 4 <= n { i + 4 } else { 0 },
        MP_BIN8 | MP_STR8 => {
            if i + 1 > n {
                return 0;
            }
            let sz = a[i] as usize;
            i += 1;
            return if sz <= n - i { i + sz } else { 0 };
        }
        MP_BIN16 | MP_STR16 => {
            if i + 2 > n {
                return 0;
            }
            let sz = read16(a, i) as usize;
            i += 2;
            return if sz <= n - i { i + sz } else { 0 };
        }
        MP_BIN32 | MP_STR32 => {
            if i + 4 > n {
                return 0;
            }
            let sz = read32(a, i) as usize;
            i += 4;
            return if sz <= n - i { i + sz } else { 0 };
        }
        MP_FIXEXT1 => return if i + 2 <= n { i + 2 } else { 0 },
        MP_FIXEXT2 => return if i + 3 <= n { i + 3 } else { 0 },
        MP_FIXEXT4 => return if i + 5 <= n { i + 5 } else { 0 },
        MP_FIXEXT8 => return if i + 9 <= n { i + 9 } else { 0 },
        MP_FIXEXT16 => return if i + 17 <= n { i + 17 } else { 0 },
        MP_EXT8 => {
            if i + 2 > n {
                return 0;
            }
            let sz = a[i] as usize;
            i += 2;
            return if sz <= n - i { i + sz } else { 0 };
        }
        MP_EXT16 => {
            if i + 3 > n {
                return 0;
            }
            let sz = read16(a, i) as usize;
            i += 3;
            return if sz <= n - i { i + sz } else { 0 };
        }
        MP_EXT32 => {
            if i + 5 > n {
                return 0;
            }
            let sz = read32(a, i) as usize;
            i += 5;
            return if sz <= n - i { i + sz } else { 0 };
        }
        _ => {}
    }

    // fixstr
    if (0xa0..=0xbf).contains(&b) {
        let sz = (b & 0x1f) as usize;
        return if sz <= n - i { i + sz } else { 0 };
    }

    // fixarray
    if (0x90..=0x9f).contains(&b) {
        let count = (b & 0x0f) as usize;
        for _ in 0..count {
            i = skip_one_d(a, n, i, depth + 1);
            if i == 0 {
                return 0;
            }
        }
        return i;
    }

    // fixmap
    if (0x80..=0x8f).contains(&b) {
        let count = (b & 0x0f) as usize;
        for _ in 0..count {
            i = skip_one_d(a, n, i, depth + 1);
            if i == 0 {
                return 0;
            }
            i = skip_one_d(a, n, i, depth + 1);
            if i == 0 {
                return 0;
            }
        }
        return i;
    }

    // array16/32
    if b == MP_ARRAY16 || b == MP_ARRAY32 {
        let count = if b == MP_ARRAY16 {
            if i + 2 > n {
                return 0;
            }
            let c = read16(a, i) as usize;
            i += 2;
            c
        } else {
            if i + 4 > n {
                return 0;
            }
            let c = read32(a, i) as usize;
            i += 4;
            c
        };
        for _ in 0..count {
            i = skip_one_d(a, n, i, depth + 1);
            if i == 0 {
                return 0;
            }
        }
        return i;
    }

    // map16/32
    if b == MP_MAP16 || b == MP_MAP32 {
        let count = if b == MP_MAP16 {
            if i + 2 > n {
                return 0;
            }
            let c = read16(a, i) as usize;
            i += 2;
            c
        } else {
            if i + 4 > n {
                return 0;
            }
            let c = read32(a, i) as usize;
            i += 4;
            c
        };
        for _ in 0..count {
            i = skip_one_d(a, n, i, depth + 1);
            if i == 0 {
                return 0;
            }
            i = skip_one_d(a, n, i, depth + 1);
            if i == 0 {
                return 0;
            }
        }
        return i;
    }

    0
}
